### PR TITLE
insorted

### PR DESCRIPTION
**Pull request by sbromberger**
Some simplification of membership testing, and some VERY modest performance improvements.

(`foo()` just removes all edges in a graph.)

Master:
```
julia> @benchmark difference(h, g)
BenchmarkTools.Trial:
  memory estimate:  40.97 MiB
  allocs estimate:  464758
  --------------
  minimum time:     100.119 ms (14.32% GC)
  median time:      126.416 ms (13.82% GC)
  mean time:        153.632 ms (25.77% GC)
  maximum time:     295.614 ms (62.45% GC)
  --------------
  samples:          34
  evals/sample:     1


julia> @benchmark foo(g)
BenchmarkTools.Trial:
  memory estimate:  4.50 MiB
  allocs estimate:  26495
  --------------
  minimum time:     21.885 ms (0.00% GC)
  median time:      27.445 ms (0.00% GC)
  mean time:        30.885 ms (3.55% GC)
  maximum time:     99.003 ms (0.00% GC)
  --------------
  samples:          162
  evals/sample:     1
```
This PR
```
julia> @benchmark difference(h, g)
BenchmarkTools.Trial:
  memory estimate:  40.97 MiB
  allocs estimate:  464758
  --------------
  minimum time:     98.077 ms (5.44% GC)
  median time:      110.199 ms (14.50% GC)
  mean time:        124.564 ms (23.34% GC)
  maximum time:     249.726 ms (58.69% GC)
  --------------
  samples:          41
  evals/sample:     1


julia> @benchmark foo(g)
BenchmarkTools.Trial:
  memory estimate:  4.50 MiB
  allocs estimate:  26495
  --------------
  minimum time:     18.158 ms (0.00% GC)
  median time:      24.558 ms (0.00% GC)
  mean time:        27.837 ms (3.87% GC)
  maximum time:     107.658 ms (0.00% GC)
  --------------
  samples:          181
  evals/sample:     1
```

### CODE REVIEW

**Comment by sbromberger** [comment]
I plan on merging this sometime next week unless there are objections cc @jpfairbanks @juliohm @somil55 

also cc @mschauer

**Comment by mschauer** [comment]
In https://github.com/JuliaGraphs/LightGraphs.jl/blob/70a42f4a94a3929ed72082e50a6cdf355d10ea38/src/graphtypes/simplegraphs/simplegraph.jl#L148 try `isempty(i)` and `first(i)`?

**Comment by sbromberger** [comment]
> try `isempty(i)` and `first(i)`?

Yes. Thanks.

File: src/graphtypes/simplegraphs/simplegraph.jl [hunk 1]
```diff
@@ -146,12 +146,12 @@ end
 
 function rem_edge!(g::SimpleGraph, e::SimpleGraphEdge)
     i = searchsorted(g.fadjlist[src(e)], dst(e))
-    length(i) > 0 || return false   # edge not in graph
-    i = i[1]
-    deleteat!(g.fadjlist[src(e)], i)
+    isempty(i) && return false   # edge not in graph
+    j = first(i)
```
Comment (mschauer) [inline src/graphtypes/simplegraphs/simplegraph.jl]: I was about to remark this.

File: src/utils.jl [hunk 2]
```diff
@@ -42,3 +42,14 @@ Unlike [`sample!`](@ref), does not produce side effects.
 sample(a::UnitRange, k::Integer; exclude = ()) = sample!(getRNG(), collect(a), k; exclude = exclude)
 
 getRNG(seed::Integer = -1) = seed >= 0 ? MersenneTwister(seed) : Base.Random.GLOBAL_RNG
+
+"""
+    insorted(x, val)
+
```
Comment (mschauer) [inline src/utils.jl]: `in` has `in(item, collection)`, contrary to LightGraphs convention `has_edge(graph, edge)` and `searchsorted(collection, item)`. I think this is fine.

File: src/graphtypes/simplegraphs/simpledigraph.jl [hunk 3]
```diff
@@ -118,9 +118,9 @@ end
 
 function rem_edge!(g::SimpleDiGraph, e::SimpleDiGraphEdge)
     has_edge(g, e) || return false
-    i = searchsorted(g.fadjlist[src(e)], dst(e))[1]
+    i = searchsortedfirst(g.fadjlist[src(e)], dst(e))
```
Comment (mschauer) [inline src/graphtypes/simplegraphs/simpledigraph.jl]: i think this does twice the work. compare https://github.com/mschauer/CausalInference.jl/blob/master/src/skeleton.jl#L11

also https://github.com/JuliaGraphs/LightGraphs.jl/blob/f435476d0d8b72ceec712e2876370c5503568e3d/src/graphtypes/simplegraphs/simplegraph.jl#L143 could check for vertex degree before searching as in has_edge

File: src/utils.jl (hunk 2 above)
Comment (sbromberger) [reply src/utils.jl]: They did that so that infixing would work, I think.

File: src/graphtypes/simplegraphs/simpledigraph.jl (hunk 3 above)
Comment (sbromberger) [reply src/graphtypes/simplegraphs/simpledigraph.jl]: It *does* do twice the work - nice catch. This was a remnant from when we had edge sets and `has_edge` was O(1). Let me fix that.

Comment (mschauer) [reply src/graphtypes/simplegraphs/simpledigraph.jl]: i like `i = searchsorted(); isempty(i) && return false; j = first(i)` more, but it should be equivalent.

File: src/utils.jl (hunk 2 above)
Comment (sbromberger) [reply src/utils.jl]: On second thought I think we should change `insorted` to be `(item, collection)`. I'll make this change.

**Comment by sbromberger** [comment]
I am going to merge this today.

**Pull request merged**